sdk/azcore/runtime/policy_retry.go

```diff
@@ -165,6 +161,19 @@ func (p *retryPolicy) Do(req *policy.Request) (resp *http.Response, err error) {
 			return
 		}
 
+		if options.ShouldRetry != nil {
```

Comment:
Does this mean that if ShouldRetry is configured by user then SDK will not check its own list of retriable errors and responsibility is passed down to application to define what all to retry?

Reply:
Users can't configure this, only SDK authors can.

The only errors that remain checked first are context errors and errors marked as NonRetriable.  There's discussion in the above comments if we want to change this so that `ShouldRetry` includes those errors as well.  Do you have an opinion?

Reply:
@vibhansa-msft the POR is to evaluate context/NonRetriable errors before the ShouldRetry handler.  Please let us know if this will be problematic for azcopy.

Reply:
Is this supposed to be inaccessible to users? RetryOptions is part of ClientOptions.

Reply:
I'm a little confused by "Users can't configure this, only SDK authors can." Since ShouldRetry is exposed in the RetryOptions, why can't a User configure this themselves?

Reply:
You're correct, I misspoke earlier.  Conceptually, I was thinking that this might make more sense to expose only to SDK authors, but that's clearly not how it's implemented in this PR.

Reply:
So we are exposing this callback which user can register while creating the client and in case of failure, callback will be hit where user application can decide whether to retry or not.
My earlier concern here was SDK has a set of retriable error codes. We are going to check those first and then hit the callback in case of no match, or if the callback is registered then we skip the SDK check and rely on user to decide when to retry and when to fail.

Reply:
When `ShouldRetry` is not `nil`, it will be called instead of checking against the configured set of HTTP status codes.  However, the retry policy will evaluate context/NonRetriable errors before the `ShouldRetry` handler is called.